### ISSUE TITLE
Offline flake development

### ISSUE THREAD

**Issue by tpwrules** [NixOS/nix#4250]
In #2890, the flake system was fixed to stop built flakes from being garbage collected. However, "nix develop" outputs still are it seems. I can't figure out how to set it up so I can develop offline if a garbage collection has run.

**Comment by edolstra** [comment]
Yeah, this has been gone for a while (1b494798360cca30971b43adda5baa154bf1991e). It will need to be revived in a way that doesn't require fetching all flake inputs eagerly.

**Comment by edolstra** [comment]
The best way to keep a `nix develop` environment around is to use a profile, e.g.

```
# nix profile --profile ./foo # create profile
# nix profile ./foo # re-enter previously created profile
```

The latter command doesn't require any fetching/building.

**Comment by tpwrules** [comment]
I discovered this independently and meant to update this issue and this indeed solves the problem, almost.

The flake inputs themselves don't seem to be considered as part of the profile (e.g. the downloaded nixpkgs tarball), so if a GC cycle is run that file has to be fetched, though nothing else has to be built and it's a quick operation.

I've worked around this by including a text file with the input paths as a build input so the inputs are considered runtime dependencies and thus kept. Then the profile truly points to everything and I can garbage collect without fear. I wonder if this could/should be done automatically in Nix itself?

**Comment by tpwrules** [comment]
I am still interested in this being fixed. But I guess my solution was equivalent to the solution that was incompatible with lazy fetching.

**Comment by mikepurvis** [comment]
I think I'm seeing this sometimes even without a GC having occurred. Literally just `nix develop` on a local, locked flake, and then a day later, `nix develop` again, and it's re-pulling the nixpkgs tarball.

Note though that this machine *is* running a Hydra instance, so possible Hydra is running GC on me?

**Comment by SuperSandro2000** [comment]
I download nixpkgs at least every day because garbage collection always collects it. Any workaround for that?

**Comment by tpwrules** [comment]
The exact implementation depends on which type of flake output you want to keep around, but [this](https://gist.github.com/tpwrules/34db43e0e2e9d0b72d30534ad2cda66d) is how I work around the problem for `nix develop`.

**Comment by SuperSandro2000** [comment]
I have also written this yesterday
```bash
gc_root_prefix="/nix/var/nix/gcroots/per-user/$USER/$(systemd-escape -p "$PWD")-flake-"
nix flake update "$@"
echo "Adding per-user gcroots..."
rm -f "$gc_root_prefix"*
nix flake archive --json 2>/dev/null | jq -r '.inputs | to_entries[] | "ln -fsT "+.value.path+" \"'"$gc_root_prefix"'"+.key+"\""' | while read -r line; do
  eval "$line"
done
```

**Comment by a-h** [comment]
I think it's impossible to do offline development at the moment due to https://github.com/NixOS/nix/issues/8953
